cpp/src/lance/arrow/dataset.cc

```diff
@@ -203,28 +214,47 @@ ::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::Make(
   if (finfo.type() == ::arrow::fs::FileType::NotFound) {
     return ::arrow::Status::IOError("Manifest not found: ", manifest_path);
   }
-  std::shared_ptr<lance::format::Manifest> manifest;
-  {
-    ARROW_ASSIGN_OR_RAISE(auto in, fs->OpenInputFile(manifest_path));
-    ARROW_ASSIGN_OR_RAISE(auto parsed, lance::format::Manifest::Parse(in, 0));
-    manifest = std::move(parsed);
-  }
+  ARROW_ASSIGN_OR_RAISE(auto manifest, OpenManifest(fs, manifest_path));
   auto impl = std::make_unique<LanceDataset::Impl>(fs, base_uri, manifest);
   return std::shared_ptr<LanceDataset>(new LanceDataset(std::move(impl)));
 }
 
+::arrow::Result<std::vector<DatasetVersion>> LanceDataset::versions() const {
+  std::vector<DatasetVersion> versions;
+  ::arrow::fs::FileSelector selector;
+  selector.base_dir = impl_->versions_dir();
+  selector.allow_not_found = true;
+  selector.recursive = false;
+
+  ARROW_ASSIGN_OR_RAISE(auto file_infos, impl_->fs->GetFileInfo(selector));
+  for (const auto& finfo : file_infos) {
+    ARROW_ASSIGN_OR_RAISE(auto manifest, OpenManifest(impl_->fs, finfo.path()));
```

Comment:
Does this need an explicit close or are these automatically cleaned up?

Reply:
this opens as `arrow::RandonInputFile` which will be closed once it is out of scope.